src/main/java/com/google/devtools/build/lib/bazel/bzlmod/GsonTypeAdapterUtil.java

```diff
@@ -82,47 +82,97 @@ public void write(JsonWriter jsonWriter, ModuleKey moduleKey) throws IOException
-          jsonWriter.value(moduleExtId.getBzlFileLabel() + "%" + moduleExtId.getExtensionName());
+          String extension = moduleExtId.getBzlFileLabel() + "%" + moduleExtId.getExtensionName();
+          if (moduleExtId.getIsolationKey().isEmpty()) {
+            jsonWriter.value(extension);
+          } else {
+            jsonWriter
+                .beginObject()
+                .name("extension")
+                .value(extension)
+                .name("isolation_key")
+                .value(moduleExtId.getIsolationKey().get().toString())
+                .endObject();
+          }
         }
 
         @Override
         public ModuleExtensionId read(JsonReader jsonReader) throws IOException {
+          if (jsonReader.peek() == JsonToken.BEGIN_OBJECT) {
+            jsonReader.beginObject();
+            Preconditions.checkState(jsonReader.nextName().equals("extension"));
+            String extensionString = jsonReader.nextString();
+            Preconditions.checkState(jsonReader.nextName().equals("isolation_key"));
+            String isolationKeyString = jsonReader.nextString();
+            jsonReader.endObject();
+            // [0] is labelString, [1] is extensionName
+            List<String> extIdParts = Splitter.on("%").splitToList(extensionString);
+            try {
+              return ModuleExtensionId.create(
+                  Label.parseCanonical(extIdParts.get(0)),
+                  extIdParts.get(1),
+                  Optional.of(ModuleExtensionId.IsolationKey.fromString(isolationKeyString)));
+            } catch (LabelSyntaxException e) {
+              throw new JsonParseException(
+                  String.format(
+                      "Unable to parse ModuleExtensionID bzl file label:  '%s' from the lockfile",
+                      extIdParts.get(0)),
+                  e);
+            } catch (ParseException e) {
+              throw new JsonParseException(
+                  String.format(
+                      "Unable to parse ModuleExtensionID isolation key:  '%s' from the lockfile",
+                      isolationKeyString),
+                  e);
+            }
+          } else {
+            String jsonString = jsonReader.nextString();
+            // [0] is labelString, [1] is extensionName
+            List<String> extIdParts = Splitter.on("%").splitToList(jsonString);
+            try {
+              return ModuleExtensionId.create(
+                  Label.parseCanonical(extIdParts.get(0)), extIdParts.get(1), Optional.empty());
+            } catch (LabelSyntaxException e) {
+              throw new JsonParseException(
+                  String.format(
+                      "Unable to parse ModuleExtensionID bzl file label:  '%s' from the lockfile",
+                      extIdParts.get(0)),
+                  e);
+            }
+          }
+        }
+      };
+
+  public static final TypeAdapter<ModuleExtensionId.IsolationKey> ISOLATION_KEY_TYPE_ADAPTER =
```

Comment:
is this ever used?

Reply:
`ModuleExtensionUsage` has a field of this type.